Add face detection constraints and VideoFrameMetadata members

ttoivone opened the pull request:
This PR supersedes previous PRs related to face detection (#57 , #48 ). It adds the constraints (and related settings and capabilities) and extends the recently introduced [VideoFrameMetadata](https://w3c.github.io/webcodecs/#dictdef-videoframemetadata) to have descriptions for faces in the frames.

The feedback has been taken into consideration, simplifying the API by removing most of the previously proposed constraints. Also the mesh-based facial description has been removed. Only those judged to be essential for good performance are left. An exception is face landmarks, which are already supported by some platforms and could be therefore immediately useful. Furthermore, HumanFace-term is used instead of more generic DetectedFace to anticipate future extensions of VideoFrameMetadata.

The PR consists of two commits. The first updates the explainer and the second updates the spec.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/ttoivone/mediacapture-extensions/pull/78.html" title="Last updated on Apr 6, 2023, 4:04 PM UTC (a1c636a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/mediacapture-extensions/78/13addf4...ttoivone:a1c636a.html" title="Last updated on Apr 6, 2023, 4:04 PM UTC (a1c636a)">Diff</a>

ttoivone commented:
Raised an [issue on Mozilla's standard positions](https://github.com/mozilla/standards-positions/issues/706)

ttoivone commented:
+@eehakkin
+@riju

chrisn commented:
Thanks! Responding having seen you post https://github.com/w3c/webcodecs/issues/607, and speaking as BBC contributor, with my Media WG chair hat removed.

We need to be sure that there are no ethical issues exposing this to the web, concerns I mentioned at the TPAC 2021 breakout meeting: https://www.w3.org/2021/10/20-webrtc-ic-minutes.html

It's good that detecting facial expressions is stated as a non-goal, but I'd recommend going further to say it "must not" rather than "does not need to". Misdetection is a concern, as mentioned in the explainer, but also, there are privacy implications of exposing inferred emotions, at least without strong user consent.

As such I'd want to see this proposal go through wide review, including Privacy and TAG.

ttoivone commented:
Changes in [5f8b11b](https://github.com/w3c/mediacapture-extensions/commit/5f8b11b827bfeaeff9970b0eae72d22fe3a5a560):
- Removed countour, replaced with with bounding-box and center-point
  for faces and landmarks, respectively
- Replaced sequence of landmarks with members in HumanFace dict
- Removed faceDetectionMaxContourPoints -- contour is now gone
- Add a separate constraint to control landmark detection
- Removed "nose" landmark to simplify -- not supported by Android/ChromeOS HAL3
- Emphasized that facial expressions must be non-goals for the spec
- Removed nullability from all members
- Specified `id` more accurately
- Once again proofread everything
- More complete acknowledgements in the explainer (send me a note if you're missing)
- Updated examples

ttoivone commented:
@youennf @jan-ivar Requesting review. I couldn't add reviewers myself for some reason.

youennf commented on face-detection-explainer.md:
```diff
@@ -12,66 +12,125 @@
 
-The field `landmarks` provides a list of facial features belonging to the detected face, such as eyes or mouth. User agent may return accurate contour for the features, but early implementations are expected to deliver only a single point corresponding to the center of a feature.
+dictionary HumanFaceLandmark {
+  Point2D           centerPoint;
```
Can we mark it required?

youennf replied:
I guess they would be initially required but in the future it could contain a bounding box as well. Is that correct?

ttoivone replied:
Correct. I did not use `required` here because in future we might have also bounding box/contour or other fields which would make `centerPoint` unnecessary. I'm happy to mark it `required` for now unless it complicates amending the spec later with other type of location information.

ttoivone replied:
After changes needed to unblock [issues 79](https://github.com/w3c/mediacapture-extensions/issues/79), this doesn't look anymore feasible.

youennf commented on face-detection-explainer.md:
```diff
@@ -12,66 +12,125 @@
 
-The field `contour` provides an arbitrary number of points which enclose the face. Sophisticated implementations could provide a large number of points but initial implementations are expected to provide four contour points located in the corners of a rectangular region which describes the face bounding box. User agent is allowed to provide a minimum of one point, in which case the point should b ...
+dictionary HumanFace {
+  long              id;
```
If zero means no tracking, we could mark id required.

ttoivone replied:
I would rather use non-existing `id` to indicate no tracking instead of a special value. It is more logical: no `id` member = no tracking. Implementations could chose whether they would include zero as one of the valid tracking `id` values or not, but Web apps would be required to handle it as a valid tracking value. Would you agree with this?

ttoivone replied:
Marking `id` as required.

youennf commented on face-detection-explainer.md:
```diff
@@ -12,66 +12,125 @@
+dictionary HumanFace {
+  long              id;
+  float             probability;
+  DOMRectReadOnly   boundingBox;
```
Can we make boundingBox required, or is there a case where we would know that there is a face in the image but we do not know where?

youennf replied:
I know see that we have a kind of union boudingBox or face landmark, so we cannot make it required.

ttoivone replied:
> we have a kind of union boudingBox or face landmark

No, if with union you mean that only either one would be valid. The intent is to allow both face `boundingBox` and face landmarks simultaneously. The `boundingBox` within the `HumanFace` dict would indicate bounding box for the whole face. Landmarks are part of human face and they can be set or not set independently of the whole face bounding box.

The reason why I did not mark `boundingBox` as `required` is the same as for the `centerPoint`: to anticipate extensions to `HumanFace` where we could have something like contour or bitmask for the face and not need bounding box anymore. If you think it is not a problem to now mark `boundingBox` as `required` and later to remove the keyword when dictionary is extended with new types of location, I am happy to add the keyword now. However, I am worried that if the member is now required, Web apps could break if the keyword is removed later and bounding box made optional. Probably this wouldn't happen though, as `humanFaceDetectionMode` would need to be set to something else than `bounding-box` too.

> is there a case where we would know that there is a face in the image but we do not know where

With this draft version, there wouldn't be such a case. However, it could be future extension: add a new `ObjectDetectionMode`  called `presence` which just indicates that a face exists in the frame without its location. Use case could be for example to prevent screensaver while user is in the front of his/her computer. We did have the `presence` mode in some earlier draft, but I removed it to simplify.

ttoivone replied:
Understood, and right, in the example there is no need to write centerPoint?.x because the metadata originates from a video track with humanFaceLandmarkDetectionMode == "center-point". 

I agree with the suggestion to document the invariants. I changed the PR as follows:

bounding-box:  This source generates metadata into {{VideoFrameMetadata}} related to human face detection including the bounding box information for each detected face. As an input, this is interpreted as a command to turn the setting of human face detection to bounding-box.

> But I would mark centerPoint as required
After changes needed to unblock [issue 79](https://github.com/w3c/mediacapture-extensions/issues/79), this doesn't look anymore feasible.

youennf commented on face-detection-explainer.md:
```diff
@@ -12,66 +12,125 @@
+enum ObjectDetectionMode {
+  "none",          // Face or landmark detection is not needed
+  "center-point",  // Center point of the detected object is returned
+  "bounding-box",  // Bounding box of the detected object is returned
```
Given the current API, it seems bounding-box is only for humanFaceDetectionMode and "center-point" only for humanFaceLandmarkDetectionMode.
Should we then have an enum for humanFaceDetectionMode and a different enum for humanFaceLandmarkDetectionMode?

youennf commented on face-detection-explainer.md:
```diff
@@ -12,66 +12,125 @@
+partially (or in special cases even fully) outside of the visible image.
+A coordinate is interpreted to represent a location in a normalized square space. The origin of
+coordinates (x,y) = (0.0, 0.0) represents the upper leftmost corner whereas the (x,y) =
+(1.0, 1.0) represents the lower rightmost corner relative to the rendered frame: the x-coordinate (columns) increases rightwards and the y-coordinate (rows) increases downwards.
```
Does it make sense to normalise?
DOMRectReadOnly has a width/height and an origin which would not be normalised but Point2D would be?
What is doing https://wicg.github.io/shape-detection-api? OS APIs?

ttoivone replied:
We had a discussion with the WebCodecs people (eg. Dan Sanders) and the issue of coordinate system turned to be a bit complicated. A normalized coordinate space seemed to be more straightforward and avoids some issues compared to natural size which doesn't always match with copyTo(). It is too bad that it doesn't seem to be consistent with the existing Web APIs. [Shape detection API](https://wicg.github.io/shape-detection-api) uses natural size (unlike we) but after the discussions it seems that the whole definition of coordinate space in that spec is not adequate, at least not for our purposes with VideoFrame which has internally a rotation, visibleRect, and codedRect. 

Therefore, I would rather stick with the definition from Image Capture spec's [Points of interest](https://www.w3.org/TR/image-capture/#points-of-interest) with normalized coordinates. I think that spec is more mainstream and avoids the above issues.

The platform APIs don't help here much, Windows uses normalized coordinates (up to INT_MAX or something) while Android/ChromeOS uses native pixel coordinates.

I think that both `Point2D` and `DOMRectReadOnly` types should use the same coordinate system for consistency. If we use normalized coordinates, also `width` and `height` should be normalized so that `right` = `x + width` for consistency. I will clarify that in the draft API once we agree on coordinate system.

ttoivone commented:
> Thanks! Responding having seen you post [w3c/webcodecs#607](https://github.com/w3c/webcodecs/issues/607), and speaking as BBC contributor, with my Media WG chair hat removed.
> 
> We need to be sure that there are no ethical issues exposing this to the web, concerns I mentioned at the TPAC 2021 breakout meeting: https://www.w3.org/2021/10/20-webrtc-ic-minutes.html
> 
> It's good that detecting facial expressions is stated as a non-goal, but I'd recommend going further to say it "must not" rather than "does not need to". Misdetection is a concern, as mentioned in the explainer, but also, there are privacy implications of exposing inferred emotions, at least without strong user consent.
> 
> As such I'd want to see this proposal go through wide review, including Privacy and TAG.

aboba commented:
Is the PR ready for CfC?

chrisn commented:
Thanks @ttoivone for updating the explainer, looks OK from my point of view.

ttoivone commented:
> Is the PR ready for CfC?

We are still waiting for review comments from WebCodecs team (Dan Sanders/Dale Curtis).

ttoivone commented:
Changes in [e2ec3d6](https://github.com/w3c/mediacapture-extensions/commit/e2ec3d659e2dcc7c33d1f8f60ce7f589b5cd6700):

- Split detection mode enum into two: HumanFaceDetectionMode and HumanFaceLandmarkDetectionMode according to feedback

Feedback was positive from WebCodecs (Dale Curtis) "Structure looks good to me for VideoFrameMetadata. I defer to @youennf  around correctness issues for what metadata should be there." @jan-ivar was removed inadvertently from the reviewer list and I still can't add reviewers myself, sorry.

@youennf @jan-ivar:  Please let us know if further updates are needed into the PR before CfC, thanks.

jan-ivar commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+if (supports.humanFaceDetectionMode) {
+  // Browser supports face contour detection.
+} else {
+  throw('Face contour detection is not supported');
+}
+
+// Open camera with face detection enabled
+const stream = await navigator.mediaDevices.getUserMedia({
+  video: { humanFaceDetectionMode: 'bounding-box' }
+});
+const [videoTrack] = stream.getVideoTracks();
+
```
I care a lot about examples, so apologies for all the comments.

I think there's a false assumption here that since `supports.humanFaceDetectionMode` is `true` here, then `videoTrack.getSettings().humanFaceDetectionMode` MUST be `'bounding-box'`.

But I don't think that's right. That would imply that the face tracking feature is implemented entirely in the browser for all camera devices without any reliance on specific device or driver hardware. Even if this is true for some user agents, I see no reason to bake this assumption into the API, when the `supports` mechanism was never meant to be where features were negotiated.

It also doesn't allow for user agents to support this for _some_ special hardware cameras but not others, nor does it appear to allow user agents to allow users to turn off these features. We should also discourage "must have" code patterns IMHO. Most likely, if face tracking cannot be gotten, then most camera apps should proceed anyway.

I think instead the model we want to promote is to apply the constraints and then read back the result with `getSettings` and react to that if necessary if so desired. e.g.
```js
if (videoTrack.getSettings().humanFaceDetectionMode != 'bounding-box') {
  throw('Face detection is not supported');
}
```
I'd prefer this added to the examples instead of using getSupportedConstraints. This also simplifies subsequent code, as I show below.

ttoivone replied:
I completely agree, and appreciate your detailed feedback. Fixing the examples.

jan-ivar commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+  const {readable} = new MediaStreamTrackProcessor({track});
+  const transformer = new TransformStream({
+    async transform(frame, controller) {
+      for (const face of frame.metadata().humanFaces || []) {
+        if (face.boundingBox) {
+          console.log(
+            `Face @ (${face.boundingBox.left}, ${face.face.boundingBox.top}), ` +
+                   `(${face.boundingBox.right}, ${face.face.boundingBox.bottom})`);
+        }
+      }
```
If the check I suggested above is added, then this can be simplified:
```suggestion
      for (const face of frame.metadata().humanFaces) {
        console.log(
          `Face @ (${face.boundingBox.left}, ${face.face.boundingBox.top}), ` +
                 `(${face.boundingBox.right}, ${face.face.boundingBox.bottom})`);
      }
```

ttoivone replied:
You suggest here two changes:
(1) assume that metadata().humanFaces always exists (so it would be empty array if no faces were detected) and
(2) for each face, assume that face.boundingBox exists because its creation is enabled in the settings

Regarding both (1) and (2):
The feedback we received and the way the metadata is implemented allows users to copy-modify and amend the metadata. Your suggested changes make assumptions which might not hold if the metadata is modified. In this example the assumptions are valid but it might cause bugs later on if the code is reused in another context. I'm mostly happy to merge these changes into the example but I think it's best to add at least a comment *why* the assumptions are made in this case (ie. metadata coming directly from the video track).

Regarding (1):
The change you suggest assumes that if there are no faces then metadata().humanFaces does exist (but is an empty array). It was not `required` in the spec proposal, so I don't see that it would be necessarily true. Maybe we should mark it as `required`?

[metadata().humanFaces was replaced with metadata().segment due to [issue 79](https://github.com/w3c/mediacapture-extensions/issues/79) but generally the same comments hold.]

jan-ivar commented on face-detection-explainer.md:
````diff
@@ -12,66 +12,129 @@
+
+The second part consists of constraints to `getUserMedia()` or `applyConstrains()` which are used to negotiate and enable the desired face detection mode. Often, camera drivers run already internally face detection for 3A, so enabling face detection might just make the results available to Web applications. The corresponding members are added also to media track settings and capabilities. 
+
+```js
+partial dictionary MediaTrackSupportedConstraints {
+  boolean humanFaceDetectionMode = true;
+  boolean humanFaceLandmarkDetectionMode = true;
+};
+
+partial dictionary MediaTrackCapabilities {
+  sequence<DOMString> humanFaceDetectionMode;
+  sequence<DOMString> humanFaceLandmarkDetectionMode;
+};
+
+partial dictionary MediaTrackConstraintSet {
+  ConstrainDOMString humanFaceDetectionMode;
+  ConstrainDOMString humanFaceLandmarkDetectionMode;
+};
+
+partial dictionary MediaTrackSettings {
+  DOMString humanFaceDetectionMode;
+  DOMString humanFaceLandmarkDetectionMode;
+};
+
+enum HumanFaceDetectionMode {
+  "none",          // Face or landmark detection is not needed
+  "bounding-box",  // Bounding box of the detected object is returned
````
WebIDL is for specs not explainers IMHO. Can this be replaced with JS snippets showing expected syntax?

if not then please replace `js` with `idl` on the first line to get the correct syntax highlighting.

alvestrand replied:
I like having WebIDL in explainers as long as we're not yet ready to land it in specs. It's a pain to have it both places.

ttoivone replied:
I think JS snippets are better for the examples at the end of the explainer. For explaining things, I found JS to be inconvenient. How would an enum be written in Javascript, for instance? Therefore, replaced js with idl.

jan-ivar commented on face-detection-explainer.md:
```diff
@@ -12,66 +12,129 @@
+### Constraints
 
-The constraint `faceDetectionMode` is used by applications to describe the level of facial data that they need. At the lowest enabled level, `presence` will return the sequence of `DetectedFace`, but the `contour` and `landmarks` sequences will be empty. When `faceDetectionMode` is `contour`, arbitrary number of points around the faces will be returned but no landmarks. An user agent might return ...
+New members are added to capabilities, constraints, and settings for Web applications to enable and control face and face landmark detection with `getUserMedia()` and `applyConstraints()` and to query capabilities of face detection with `getCapabilities()` methods. Web applications should not ask more facial metadata than what they need to limit computation. For example, if an applications is con ...
```
The sentence _"Web applications should not ask more facial metadata than what they need to limit computation"_ has a typo, but also seems to be admonishing web developers to behave a certain way, which reads a bit odd to me.

Could it be rephrased to emphasize the benefits of asking for less information?

ttoivone replied:
Rephrased to "Web applications can reduce the computational load on the user agent by only requesting the necessary amount of facial data, rather than asking for more than they need."

jan-ivar commented on face-detection-explainer.md:
```diff
@@ -12,66 +12,129 @@
+New members are added to capabilities, constraints, and settings for Web applications to enable and control face and face landmark detection with `getUserMedia()` and `applyConstraints()` and to query capabilities of face detection with `getCapabilities()` methods. Web applications should not ask more facial metadata than what they need to limit computation. For example, if an applications is con ...
 
-At the highest level, when `faceDetectionMode` is `landmarks`, the full precision contour which is available is returned along with landmark features.
+The enumeration constraints `humanFaceDetectionMode` and `humanFaceLandmarkDetectionMode` set the level of detection needed for human faces and their landmarks, respectively. These settings can be one of the enumeration values in `HumanFaceDetectionMode` and `HumanFaceLandmarkDetectionMode`. When `humanFaceDetectionMode` is `"bounding-box"`, user agent must attempt face detection and set the meta ...
```
"enumeration constraints" is an odd term. They are DOMString constraints. Maybe just say "constraints"?

Also to nitpick on language, these constraints are "applied" not "set". And even when successfully set, it is probably wise for the application to read back `videoTrack.getSettings()` to verify that the constraints actually stuck, as I believe that is the only point at which user agents MUST perform the feature (e.g. a user COULD have configured their user agent to not expose this functionality, even though the constraint name is recognized by `getSupportedConstraints`).

ttoivone replied:
Removed "enumeration", rephrares the paragraph. It now begins with "The constraints `humanFaceDetectionMode` and `humanFaceLandmarkDetectionMode` request the metadata detail level that the application wants for human faces and their landmarks, respectively."

jan-ivar commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+            frames originating from the same {{MediaStreamTrack}} source, {{id}} is set to the same integer value
+            for the face in all frames.</p>
+            <p>User agent MUST NOT select the assigned value of {{id}} in such a way that the detected faces could
+            be correlated to match in any way between different {{MediaStreamTrack}} objects.</p>
```
This means that a `const trackB = trackA.clone()` will have different ids. Was that the intent?

ttoivone replied:
Yes, it is the main rule and I don't see a reason to make an exception in this case. Do you think there would be a reason to do otherwise?

ttoivone replied:
I noticed that actually elsewhere in the spec and explainer notes of different {{MediaStreamTrack}} **sources** are made, contradicting with this sentence. Therefore, changing this sentence to: "_be correlated to match between different {{MediaStreamTrack}} sources._"

jan-ivar commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+    customFaceDetection = true;
+  }
+  if ((capabilities.eyeGazeCorrection || []).includes(true)) {
+    // The platform supports eye gaze correction.
+    // Let's use platform eye gaze correction and skip the custom one.
+    await videoTrack.applyConstraints({
+      advanced: [{eyeGazeCorrection: true}]
+    });
```
Please remove mention of gaze correction from examples, as it's surprising here (is it from an earlier version of the PR?)

ttoivone replied:
Removed.

jan-ivar reviewed (changes requested):
Hi all, thanks so much for all this work! Overall I think it looks good, just lots of nits and comments since it's taken me a while to dive in and review this. My apologies for that.

To summarize some of my feedback:
- document invariants better around reading metadata, so web devs don't have to `?.` every data point
- clarify, maybe even simplify `probability`?
- orient examples around detecting support using getSettings() not getSupportedConstraints()
- define what left/right mean (eye of human or beholder?)
- clarify if the two constraints are dependent or not (and why there are two)
- remove gaze correction

Thanks! Happy to take changes after CfC has concluded.

martinthomson commented on face-detection-explainer.md:
```diff
@@ -12,66 +12,129 @@
-The field `contour` provides an arbitrary number of points which enclose the face. Sophisticated implementations could provide a large number of points but initial implementations are expected to provide four contour points located in the corners of a rectangular region which describes the face bounding box. User agent is allowed to provide a minimum of one point, in which case the point should b ...
+dictionary HumanFace {
+  long              id;
+  float             probability;
```
This is probably better expressed as a "confidence".

It's not really a probability: the face is either there or not.  This value is intended to represent how confident the API is that the face is there.

aboba replied:
I assume that here "probability" is the estimated probability provided by the model, based on the data it was trained on.

Confidence is typically not expressed as a single number, but rather a range.

jan-ivar replied:
> I assume that here "probability" is the estimated probability provided by the model,

Naming aside, leaving the range of possible values implementation-specific may make it hard for apps to reason about the values. E.g. if one implementation's model rarely ever outputs above 0.7 for some technical purity reason, even in a well-lit face front situation, and another model reliably produces 1.0 except in corner cases, it could lead to weird web compat issues.

Are there precedents we could look at? Would it make sense to say something like `1.0` SHOULD represent the highest value the model can consistently output?

dontcallmedom replied:
I think confidence is the right term.
> Each confidence estimate might be seen as a probability of correctness, reflect-
ing the model’s uncertainty about a detection or the pixel mask. During inference,
we expect the estimated confidence to match the observed precision for a predic-
tion. For example, given 100 predictions with 80% confidence each, we expect 80
predictions to be correctly predicted
https://arxiv.org/pdf/2202.12785

That very paper describes how to ensure proper calibration of this confidence for object detection; I'm not sure whether we want to go to that level of specificity though.

ttoivone replied:
> leaving the range of possible values implementation-defined may make it hard for apps to reason about the values.

I agree with this 100% and think that unless the value is properly defined it shouldn't be included into the spec.

The paper that dom quotes says "confidence estimate **might be seen** as a probability of correctness".  To me this doesn't look like very rigorous or widely used definition. I'd rather go to commonly used definitions, such as [Wikipedia](https://en.wikipedia.org/wiki/Probability). I believe what we would like to have here is conditional probability, but the condition needs to be more precisely defined than what it used to be. In the latest explainer/spec draft I changed the definition as follows:

"the estimate of the conditional probability that the segmented object actually is of the type indicated by the {{Segment/type}} member on the condition that the detection has been made." (where {{Segment/type}} describes whether it is a face or some other type of object; see [issue 79](https://github.com/w3c/mediacapture-extensions/issues/79).

It is an estimate, not exact probability. As an implementor advise, the implementor could run the face detection algorithm on some test videos and of the detections that the algorithm made, check how many are correct (missed faces would not affect the conditional probability here although the algorithm quality affects that too). Other factors could also be taken into account, such as confidence/score from the algorithm, camera image quality, etc.

chrisn commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+          <dt><dfn><code>probability</code></dfn> of type <span class="idlMemberType">{{float}}</span></dt>
+          <dd>
+            <p>If set, a probability value in range (0,1].
+            The approximate probability of the {{HumanFace}} specifying an actual human face.</p>
```
In addition, range notation can be confusing to read, so I'd suggest explaining the range of allowed values in full: If set, a probability value greater than zero and with a maximum of 1.

ttoivone replied:
@jan-ivar By the commonly used definition of probability, it is always zero or greater with a maximum of one. Since it doesn't make sense to return zero as a probability, it could be used as a special value.

I updated the specific definition of `probability` in the spec, hopefully it is more precise now. Quoting the latest version:
 "[probability] is the estimate of the conditional probability that the segmented object actually is of the type indicated by the {{Segment/type}} member on the condition that the detection has been made." Note that it would be an estimate of the probability, not probability itself which is usually impossible to exactly calculate.

An implementer could estimate the probability by feeding some test video into the face detection algorithm and manually verifying how large percentage of the detections are correct (if the algorithm outputs a "score", "confidence" or similar, this could be taken also into account). 

A probability estimate of `1` would likely be inaccurate for non-synthetic video, but I think for a highly accurate algorithm it might make sense to return it (close enough to `1`). I think generally `1` definitely should not be a common value. Typically platform face detection implementation are not that accurate. 

@chrisn Changed the range notation.

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
   </section>
+  <section>
+    <h2>Human faces</h2>
+    <p>Human face metadata describes the human faces in video frames. It can
```
Changed to "Human face metadata describes the geometrical information of human faces". I think that is more precise. "presence" sounds like boolean to me, I would understand it describing whether there are faces on the face or not.

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+    <h2>Human faces</h2>
+    <p>Human face metadata describes the human faces in video frames. It can
+    be set by web applications using the standard means when creating
+    {{VideoFrameMetadata}} for {{VideoFrame}}s or it can be set by an user agent
```
Fixed (and search-replaced for other similar typos)

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+    when the media track constraints, defined below, are used to
+    enable face detection for the {{MediaStreamTrack}} which provides the
+    {{VideoFrame}}s.</p>
+    <p>The facial metadata may be used by video encoders to enhance the quality
```
Fixed.

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+        <dl class="dictionary-members" data-dfn-for="HumanFace" data-link-for="HumanFace">
+          <dt><dfn><code>id</code></dfn> of type <span class="idlMemberType">{{long}}</span></dt>
+          <dd>
+            <p>If set, an unique identifer of a face within a sequence. If the same face can be tracked in multiple
```
As per issue [Segmentation metadata](https://github.com/w3c/mediacapture-extensions/issues/79) we now describe segments. I changed this correspondingly to "An identifier of the segment, unique within a sequence. ".
Also removed "if set" since it's now required.

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+          <dt><dfn><code>id</code></dfn> of type <span class="idlMemberType">{{long}}</span></dt>
+          <dd>
+            <p>If set, an unique identifer of a face within a sequence. If the same face can be tracked in multiple
+            frames originating from the same {{MediaStreamTrack}} source, {{id}} is set to the same integer value
+            for the face in all frames.</p>
```
Changed as you suggested. Since now we allow tracking generic objects, not just faces, it is now: "the user agent SHOULD use the same {{Segment/id}}  for the object in all frames."

ttoivone commented on face-detection-explainer.md:
```diff
@@ -12,66 +12,129 @@
-At the highest level, when `faceDetectionMode` is `landmarks`, the full precision contour which is available is returned along with landmark features.
+The enumeration constraints `humanFaceDetectionMode` and `humanFaceLandmarkDetectionMode` set the level of detection needed for human faces and their landmarks, respectively. These settings can be one of the enumeration values in `HumanFaceDetectionMode` and `HumanFaceLandmarkDetectionMode`. When `humanFaceDetectionMode` is `"bounding-box"`, user agent must attempt face detection and set the meta ...
 
+If face detection mode is `"none"`, also landmark detection setting must be the same `"none"` because without face detection results the landmark data can not be set either.
```
After changing the metadata according to this [issue](https://github.com/w3c/mediacapture-extensions/issues/79) landmark metadata is not anymore syntactically tied to facial metadata and could therefore stored independently. Earlier, facial landmarks were part of the facial metadata and if the latter was not there, the prior couldn't have been either. 

However, even after the metadata syntax change, platform APIs typically don't allow detecting landmarks without also detecting faces. Also I think it would rarely be useful anyway. I think the best is to leave it to the user agent to decide if there is a dependency between the settings `humanFaceDetectionMode` and `humanFaceLandmarkDetectionMode`. I can see here two reasonable behaviors: 1) When `humanFaceLandmarkDetectionMode` is constrained to be `"center-point"`, `humanFaceDetectionMode` is set implicitly to `"bounding-box"` or 2) they can be set independently.

"Landmarks are part of human face and they can be set or not set independently of the whole face bounding box." -- this was inaccurate and misleading. It would have been more correct to say that web application may disable or enable facial landmark detection as long as face metadata is enabled. After the metadata syntax change, the sentence is even less true so changing it into following:

"On many platform APIs, enabling face landmark detection implies also enabling face detection. Therefore the user agent could enable face detection (setting it to `"bounding-box"`) when a web application applies the constraint to enable facial landmark detection. "

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+          </dd>
+          <dt><dfn><code>probability</code></dfn> of type <span class="idlMemberType">{{float}}</span></dt>
+          <dd>
+            <p>If set, a probability value in range (0,1].
```
I changed it to `required probability` and range between [0,1]. I would suggest 0 to mean that the probability estimate is not supported or known (more of this below). 0 is not otherwise a valid probability for a detected face, as you say.

In the practical cases I can think of, if user agent could assign `probability` for a face, it would assign some probability also to all other faces. This might not be true between different types of objects, ie., if faces have (nonzero) probability, landmarks might not have any probability estimate (ie. it would be set to zero).

> If the implementation isn't sharing probability at all, then an app would seem to have no choice but to assume that it is 1

I don't agree with this. I think there's a huge difference whether a probability for an event is known or calculated and whether it is not. A web application might want to run a secondary, more accurate detection algorithm for those face candidates which have low or non-existing (zero) probabilities.

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+          </dd>
+          <dt><dfn><code>leftEye</code></dfn> of type <span class="idlMemberType"><code>{{HumanFaceLandmark}}</code></span></dt>
+          <dd>
+            <p>If exists, describes the face left eye properties.</p>
```
Thanks, note taken. After changing most members to `required`, not needed for now.

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+          </dd>
+          <dt><dfn><code>rightEye</code></dfn> of type <span class="idlMemberType"><code>{{HumanFaceLandmark}}</code></span></dt>
+          <dd>
+            <p>If exists, describes the face right eye properties.</p>
```
Silly me, I thought it was obvious :-) The terms oculus sinister and oculus dexter have well-defined meanings. Maybe we could use those? Updated the spec and explainer.

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+          <dt><dfn><code>humanFaceDetectionMode</code></dfn> of type <span class="idlMemberType">{{DOMString}}</span></dt>
+          <dd>
+            <p>Current <a>face detection mode</a> setting.
+            The string MUST be one of the members of {{HumanFaceDetectionMode}}.</p>
```
Fixed.

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+        <dl class="dictionary-members" data-dfn-for="MediaTrackConstraintSet" data-link-for="MediaTrackConstraintSet">
+          <dt><dfn><code>humanFaceDetectionMode</code></dfn> of type <span class="idlMemberType">{{ConstrainDOMString}}</span></dt>
+          <dd>
+            <p>The string MUST be one of the members of {{HumanFaceDetectionMode}}.
```
Fixed for both `humanFaceDetectionMode` and `humanFaceLandmarkDetectionMode`.

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+          <dt><dfn><code>none</code></dfn></dt>
+          <dd>
+            <p>This source does not offer object detection for human faces.
+            For setting, this is interpreted as a command to turn off
```
Fixed, again for both `humanFaceDetectionMode` and `humanFaceLandmarkDetectionMode`.

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+          <dt><dfn><code>none</code></dfn></dt>
+          <dd>
+            <p>This source does not offer object detection for human face landmarks.
+            For setting, this is interpreted as a command to turn off
```
Fixed.

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+// Check if face detection is supported by the browser
+const supports = navigator.mediaDevices.getSupportedConstraints();
+if (supports.humanFaceDetectionMode) {
+  // Browser supports face contour detection.
```
Removed this from examples.

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+    // The platform supports background blurring.
+    // Let's use platform background blurring and skip the custom one.
+    await track.applyConstraints({
+      advanced: [{backgroundBlur: capabilities.backgroundBlur.max}]
```
Removed, these were not necessary.

youennf commented:
Given @ttoivone comment, I think we should review the PR at next editor's meeting.

youennf commented on index.html:
```diff
@@ -654,5 +654,368 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+            frames originating from the same {{MediaStreamTrack}} source, the user agent SHOULD use the same {{Segment/id}}
+            for the segment in all frames. {{Segment/id}} is also used in conjunction with the member {{Segment/partOf}}.</p>
+            <p>The user agent MUST NOT select the value to assign to {{Segment/id}} in such a way that the detected segments
+            could be correlated to match between different {{MediaStreamTrack}} sources.</p>
```
What about MediaStreamTrack clones?

ttoivone replied:
My thinking was that any MediaStreamTrack clones have "frames originating from the same {{MediaStreamTrack}} source". Is that not true? I will clarify the PR and mention MediaStreamTrack clones explicitly.

youennf commented on index.html:
```diff
@@ -654,5 +654,368 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+      <pre class="idl">
+dictionary Segment {
+  required SegmentType type;
+  required long        id;
```
The requirement is probably for the segments sequence to not contain two segment dictionaries with the same id.
Somehow, it seems the segments should be map-like (id -> Segment) instead of sequence.

ttoivone replied:
Yes, it is the requirement. But it looks that maplike is for interfaces, not dictionaries. Changing to interface might not be possible, because this must be part of the VideoFrameMetadata dictionary. Also is maplike even serializable?

youennf commented on index.html:
```diff
@@ -654,5 +654,368 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+dictionary Segment {
+  required SegmentType type;
+  required long        id;
+  required long        partOf;
```
I see if 0, then it is not part of any root element.
undefined might be better.
Also this would mean 0 should not be used as id.

ttoivone replied:
I assume your note relates only to the `partOf` member. I agree. I'll remove `required` here and undefined would then indicate that the segment is not part of any other segment.

ttoivone replied:
It seems not logical to me that `partOf` is not required (because it is not always needed) but `probability` is required (and a special value indicates that it is not given). Shouldn't both be either be required (with a special value) or not required?

ttoivone commented on index.html:
```diff
@@ -654,5 +654,400 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+      <h3>{{MediaTrackConstraintSet}}</h3>
+<pre class="idl">
+partial dictionary MediaTrackConstraintSet {
+  ConstrainDOMString humanFaceDetectionMode;
+  ConstrainDOMString humanFaceLandmarkDetectionMode;
```
@jan-ivar As per the meeting Mar 30, changing the constraint as you suggest.

ttoivone commented on index.html:
```diff
@@ -654,5 +654,368 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+      <h3>{{VideoFrameMetadata}}</h3>
+      <pre class="idl">
+partial dictionary VideoFrameMetadata {
+  sequence&lt;Segment&gt; segment;
```
Changed.

ttoivone commented:
Changes in the latest [update](https://github.com/w3c/mediacapture-extensions/compare/67c9dba6fb24f3efbf03a6d9f76b2ff098e5591c..311cd1cd34005730d2da27b07eb2c9d82e7ed41c) of the PR:

- Removed a constraint, only one remaining
- partOf no more required
- Explicitly specify that cloned MediaStreamTracks should have the same `id`s
- Many typo fixes and refactoring the text

jan-ivar commented on index.html:
```diff
@@ -654,5 +654,327 @@ <h2>Exposing change of MediaStreamTrack configuration</h2>
+  long                 partOf;
+  required float       probability;
+  Point2D              centerPoint;
+  DOMRectReadOnly      boundingBox;
```
This uses what to me seems like a convenience interface inside a dictionary, which seems a bit unusual, and might be odd since VideoFrameMetadata might be used as input to the VideoFrame constructor.

Suggest using DOMRectInit instead.

youennf replied:
Looking at shape detection API, they went with DOMRectReadOnly.
I'll file an issue to see how we should handle the situation here, it seems we should converge somehow between the two APIs or at least know which model is the reference for future developments.

jan-ivar commented:
Editors agreed to merge with the change above.
